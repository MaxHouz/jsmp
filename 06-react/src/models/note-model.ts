export class NoteModel {
    constructor(
        public title: string,
        public text: string,
        public date: any,
        public done: boolean,
        public archieved: boolean,
        public id?: number
    ){}
}