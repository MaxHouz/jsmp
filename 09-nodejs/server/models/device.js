const mongoose = require('mongoose');

const Device = mongoose.model('Device', {
    name: String,
    address: String,
    port: Number,
    state: String,
    groupId: String,
    log: [{
        date: String,
        action: String
    }]
});

module.exports = Device;
