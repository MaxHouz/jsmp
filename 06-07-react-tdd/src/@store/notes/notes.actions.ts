import { NoteModel } from '../../models/note-model';
import { NotesState } from './notes.state';
import * as notesApi from '../../api/notes-api';
import * as NotesActionTypes from './notes.action-types';
import { ThunkAction } from 'redux-thunk';
import { Action, Dispatch } from 'redux';


export function getNotesList(): ThunkAction<Promise<Action>, NotesState, void, Action> {
    return async (dispatch: Dispatch): Promise<Action> => {
        try {
            const res = await notesApi.getNotesList();

            return dispatch(getNotesListSuccess(res.data));
        } catch {
            return dispatch(getNotesListError());
        }
    }
}

export function getNotesListSuccess(payload: NoteModel[]): NotesActionTypes.GetNotesListSuccess {
    return {
        type: NotesActionTypes.NotesActions.GET_NOTES_LIST_SUCCESS,
        payload
    }
}

export function getNotesListError(): NotesActionTypes.GetNotesListError {
    return {
        type: NotesActionTypes.NotesActions.GET_NOTES_LIST_ERROR,
        payload: 'Notes loading failed'
    }
}

export function addNote(note: NoteModel): ThunkAction<Promise<Action>, NotesState, void, Action> {
    return async (dispatch: Dispatch): Promise<Action> => {
        try {
            const res = await notesApi.addNote(note);

            return dispatch(addNoteSuccess(res.data));
        } catch {
            return dispatch(addNoteError());
        }
    }
}

export function addNoteSuccess(payload: NoteModel): NotesActionTypes.AddNoteSuccess {
    return {
        type: NotesActionTypes.NotesActions.ADD_NOTE_SUCCESS,
        payload
    }
}

export function addNoteError(): NotesActionTypes.AddNoteError {
    return {
        type: NotesActionTypes.NotesActions.ADD_NOTE_ERROR,
        payload: 'Adding Note Failed'
    }
}

export function updateNote(note: NoteModel): ThunkAction<Promise<Action>, NotesState, void, Action> {
    return async (dispatch: Dispatch): Promise<Action> => {
        try {
            const res = await notesApi.updateNote(note);

            return dispatch(updateNoteSuccess(res.data));
        } catch {
            return dispatch(updateNoteError());
        }
    }
}

export function updateNoteSuccess(payload: NoteModel): NotesActionTypes.UpdateNoteSuccess {
    return {
        type: NotesActionTypes.NotesActions.UDPATE_NOTE_SUCCESS,
        payload
    }
}

export function updateNoteError(): NotesActionTypes.UpdateNoteError {
    return {
        type: NotesActionTypes.NotesActions.UPDATE_NOTE_ERROR,
        payload: 'Note Update Failed'
    }
}

export function deleteNote(id: number): ThunkAction<Promise<Action>, NotesState, void, Action> {
    return async (dispatch: Dispatch): Promise<Action> => {
        try {
            await notesApi.deleteNote(id);

            return dispatch(deleteNoteSuccess(id));
        } catch {
            return dispatch(deleteNoteError());
        }
    }
}

export function deleteNoteSuccess(payload: number): NotesActionTypes.DeleteNoteSuccess {
    return {
        type: NotesActionTypes.NotesActions.DELETE_NOTE_SUCCESS,
        payload
    }
}

export function deleteNoteError(): NotesActionTypes.DeleteNoteError {
    return {
        type: NotesActionTypes.NotesActions.DELETE_NOTE_ERROR,
        payload: 'Note Delete Failed'
    }
}

export function search(payload: string): NotesActionTypes.Search {
    return {
        type: NotesActionTypes.NotesActions.SEARCH,
        payload
    }
}