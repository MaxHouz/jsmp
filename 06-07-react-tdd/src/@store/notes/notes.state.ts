import { NoteModel } from '../../models/note-model';

export interface NotesState {
    notes: NoteModel[],
    error: string | null,
    search: string
}

export const NotesInitialState: NotesState = {
    notes: [],
    error: null,
    search: ''
}
