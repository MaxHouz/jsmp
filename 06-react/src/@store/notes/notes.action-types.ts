import { NoteModel } from '../../models/note-model';
import { Action } from 'redux';

export enum NotesActions {
    GET_NOTES_LIST = '[Notes] GET_NOTES_LIST',
    GET_NOTES_LIST_SUCCESS = '[Notes] GET_NOTES_LIST_SUCCESS',
    GET_NOTES_LIST_ERROR = '[Notes] GET_NOTES_LIST_ERROR',

    ADD_NOTE = '[Notes] ADD_NOTE',
    ADD_NOTE_SUCCESS = '[Notes] ADD_NOTE_SUCCES',
    ADD_NOTE_ERROR = '[Notes] ADD_NOTE_ERROR',

    UPDATE_NOTE = '[Notes] UPDATE_NOTE',
    UDPATE_NOTE_SUCCESS = '[Notes] UDPATE_NOTE_SUCCESS',
    UPDATE_NOTE_ERROR = '[Notes] UPDATE_NOTE_ERROR',

    DELETE_NOTE = '[Notes] DELETE_NOTE',
    DELETE_NOTE_SUCCESS = '[Notes] DELETE_NOTE_SUCCESS',
    DELETE_NOTE_ERROR = '[Notes] DELETE_NOTE_ERROR',

    SEARCH = '[Notes] SEARCH'
}

export interface GetNotesList extends Action<NotesActions.GET_NOTES_LIST> {
    type: NotesActions.GET_NOTES_LIST;
}

export interface GetNotesListSuccess extends Action<NotesActions.GET_NOTES_LIST_SUCCESS> {
    type: NotesActions.GET_NOTES_LIST_SUCCESS;
    payload: NoteModel[];
}

export interface GetNotesListError extends Action<NotesActions.GET_NOTES_LIST_ERROR> {
    type: NotesActions.GET_NOTES_LIST_ERROR;
    payload: string;
}

export interface AddNote extends Action<NotesActions.ADD_NOTE> {
    type: NotesActions.ADD_NOTE;
    payload: NoteModel;
}

export interface AddNoteSuccess extends Action<NotesActions.ADD_NOTE_SUCCESS> {
    type: NotesActions.ADD_NOTE_SUCCESS;
    payload: NoteModel;
}

export interface AddNoteError extends Action<NotesActions.ADD_NOTE_ERROR> {
    type: NotesActions.ADD_NOTE_ERROR;
    payload: string;
}

export interface UpdateNote extends Action<NotesActions.UPDATE_NOTE> {
    type: NotesActions.UPDATE_NOTE;
    payload: NoteModel;
}

export interface UpdateNoteSuccess extends Action<NotesActions.UDPATE_NOTE_SUCCESS> {
    type: NotesActions.UDPATE_NOTE_SUCCESS;
    payload: NoteModel;
}

export interface UpdateNoteError extends Action<NotesActions.UPDATE_NOTE_ERROR> {
    type: NotesActions.UPDATE_NOTE_ERROR;
    payload: string;
}

export interface DeleteNote extends Action<NotesActions.DELETE_NOTE> {
    type: NotesActions.DELETE_NOTE;
    payload: number;
}

export interface DeleteNoteSuccess extends Action<NotesActions.DELETE_NOTE_SUCCESS> {
    type: NotesActions.DELETE_NOTE_SUCCESS;
    payload: number;
}

export interface DeleteNoteError extends Action<NotesActions.DELETE_NOTE_ERROR> {
    type: NotesActions.DELETE_NOTE_ERROR;
    payload: string;
}

export interface Search extends Action<NotesActions.SEARCH> {
    type: NotesActions.SEARCH;
    payload: string;
}

export type NoteActionsType = GetNotesList
 | GetNotesListSuccess
 | GetNotesListError
 | AddNote
 | AddNoteSuccess
 | AddNoteError
 | UpdateNote
 | UpdateNoteSuccess
 | UpdateNoteError
 | DeleteNote
 | DeleteNoteSuccess
 | DeleteNoteError
 | Search;