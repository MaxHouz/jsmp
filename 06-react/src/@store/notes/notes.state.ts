import { NoteModel } from '../../models/note-model';

export interface NotesState {
    notes: NoteModel[],
    error: string | null
}

export const NotesInitialState: NotesState = {
    notes: [],
    error: null
}
