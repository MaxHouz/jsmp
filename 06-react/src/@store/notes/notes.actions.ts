import { NoteModel } from '../../models/note-model';
import { NotesState } from './notes.state';
import * as notesApi from '../../api/notes-api';
import * as NotesActionTypes from './notes.action-types';
import { ThunkAction } from 'redux-thunk';
import { Action, ActionCreator, Dispatch } from 'redux';


export function getNotesList(): ThunkAction<Promise<Action>, NotesState, void, Action> {
    return async (dispatch: Dispatch): Promise<Action> => {
        try {
            const res = await notesApi.getNotesList();

            return dispatch(getNotesListSuccess(res.data));
        } catch {
            return dispatch(getNotesListError())
        }
    }
}

export function getNotesListSuccess(payload: NoteModel[]): NotesActionTypes.GetNotesListSuccess {
    return {
        type: NotesActionTypes.NotesActions.GET_NOTES_LIST_SUCCESS,
        payload
    }
}

export function getNotesListError(): NotesActionTypes.GetNotesListError {
    return {
        type: NotesActionTypes.NotesActions.GET_NOTES_LIST_ERROR,
        payload: 'Notes loading failed'
    }
}