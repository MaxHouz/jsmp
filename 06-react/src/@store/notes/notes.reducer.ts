import { NotesState, NotesInitialState } from "./notes.state";
import { NoteActionsType, NotesActions } from "./notes.action-types";

export function NotesReducer(state: NotesState = NotesInitialState, action: NoteActionsType): NotesState {
    switch(action.type) {
        case NotesActions.GET_NOTES_LIST_SUCCESS: {
            return {
                ...state,
                notes: action.payload,
                error: null
            }
        }
        case NotesActions.GET_NOTES_LIST_ERROR:
        case NotesActions.ADD_NOTE_ERROR:
        case NotesActions.UPDATE_NOTE_ERROR:
        case NotesActions.DELETE_NOTE_ERROR: {
            return {
                ...state,
                error: action.payload
            }
        }
        case NotesActions.ADD_NOTE_SUCCESS: {
            state.notes.push(action.payload);

            return {
                ...state,
                notes: [...state.notes],
                error: null
            }
        }
        case NotesActions.UDPATE_NOTE_SUCCESS: {
            const updatedNote = action.payload;
            const index = state.notes.findIndex(note => note.id === updatedNote.id);
            state.notes.splice(index, 1, updatedNote);
            return {
                ...state,
                notes: [...state.notes],
                error: null
            }
        }
        case NotesActions.DELETE_NOTE_SUCCESS: {
            const index = state.notes.findIndex(note => note.id === action.payload);
            state.notes.splice(index, 1);
            return {
                ...state,
                notes: [...state.notes],
                error: null
            }
        }
        default: {
            return state;
        }
    }
}