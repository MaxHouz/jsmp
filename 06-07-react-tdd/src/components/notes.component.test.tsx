import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { NotesComponent, mapStateToProps, mapDispatchToProps } from './notes.component';
import { NotesState } from '../@store/notes/notes.state';
import { NoteModel } from '../models/note-model';

configure({adapter: new Adapter()});
describe('NotesComponent', () => {

     let component: any;
     let res, expRes;
     const mockNotes = [{
        title: 'mockTitleValue',
        text: 'mockTextValue',
        done: false,
        archieved: false,
        id: 3
     }]
     const searckMock='mock';
     const updateNoteMock = jest.fn();
     const deleteNoteMock = jest.fn();

     beforeEach(() => {
         component = shallow(
             <NotesComponent
                notes={mockNotes}
                error={null}
                search={searckMock}
                updateNote={updateNoteMock}
                deleteNote={deleteNoteMock}
             />
         )
     })

     it('should render component', () => {
        expect(component).toMatchSnapshot();
    })

    it('should update note state on complete', () => {
        component.find('NoteItemComponent').prop('onNoteComplete')(mockNotes[0]);
        expect(updateNoteMock).toHaveBeenCalledWith({
            ...mockNotes[0], 
            done: true
        })
    })

    it('should update note state on uncomplete', () => {
        component.find('NoteItemComponent').prop('onNoteUncomplete')(mockNotes[0]);
        expect(updateNoteMock).toHaveBeenCalledWith({
            ...mockNotes[0], 
            done: false
        })
    })

    it('should update note state on archieve', () => {
        component.find('NoteItemComponent').prop('onNoteArchieve')(mockNotes[0]);
        expect(updateNoteMock).toHaveBeenCalledWith({
            ...mockNotes[0], 
            archieved: true
        })
    })

    it('should delete note', () => {
        component.find('NoteItemComponent').prop('onNoteDelete')(10);
        expect(deleteNoteMock).toHaveBeenCalledWith(10)
    })

    it('should not dearchieve note', () => {
        const res = component.find('NoteItemComponent').prop('onNoteDearchieve')();
        expect(res).toEqual(undefined);
    })

    it('should map state to props', () => {
        const mockState = {
            notes: [],
            error: null,
            search: 'mockSearch'
        }
        
        expRes = {
            notes: mockState.notes,
            error: mockState.error,
            search: mockState.search
        }
        res = mapStateToProps(mockState as NotesState);

        expect(res).toEqual(expRes);
    })

    it('should not show archieved notes', () => {
        const mockArchievedNotes = [{
            title: 'mockTitleValue',
            text: 'mockTextValue',
            done: false,
            archieved: true,
            id: 3
         }]

         const component = shallow(
            <NotesComponent
               notes={mockArchievedNotes}
               error={null}
               search={searckMock}
               updateNote={updateNoteMock}
               deleteNote={deleteNoteMock}
            />
        )

        res = component.find('NoteItemComponent').exists();

        expect(res).toBeFalsy;
    })

    it('should not show notes that do not match search query', () => {
         const component = shallow(
            <NotesComponent
               notes={mockNotes}
               error={null}
               search={'randomQuery'}
               updateNote={updateNoteMock}
               deleteNote={deleteNoteMock}
            />
        )

        res = component.find('NoteItemComponent').exists();

        expect(res).toBeFalsy;
    })

    it('should map updateNote dispatch to props', () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).updateNote({} as NoteModel);
        
        expect(typeof dispatch.mock.calls[0][0]).toBe('function');
    })

    it('should map deleteNote dispatch to props', () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).deleteNote(3);
        
        expect(typeof dispatch.mock.calls[0][0]).toBe('function');
    })

})