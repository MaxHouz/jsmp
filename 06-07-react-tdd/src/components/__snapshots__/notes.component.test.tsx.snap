// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesComponent should render component 1`] = `ShallowWrapper {}`;
